File: VrchAndroid/app/src/main/java/nokamoto/github/com/vrchandroid/firebase/FirebaseMessage.java
package nokamoto.github.com.vrchandroid.firebase;

import android.util.Log;

import com.google.common.base.Optional;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.ServerValue;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import nokamoto.github.com.vrchandroid.main.Platform;
import nokamoto.github.com.vrchandroid.main.WhoAmI;

public class FirebaseMessage {
    private final static String TAG = FirebaseMessage.class.getSimpleName();

    private WhoAmI who;
    private String displayName;
    private String uid;
    private String message;
    private String uuid;
    private long createdAt;
    private Platform platform;

    private final static String WHO = "who";
    private final static String DISPLAY_NAME = "display_name";
    private final static String UID = "uid";
    private final static String MESSAGE = "message";
    private final static String UUID = "uuid";
    final static String CREATED_AT = "created_at";
    private final static String PLATFORM = "platform";

    private FirebaseMessage(WhoAmI who, String displayName, String uid, String message, String uuid, long createdAt, Platform platform) {
        this.who = who;
        this.displayName = displayName;
        this.uid = uid;
        this.message = message;
        this.uuid = uuid;
        this.createdAt = createdAt;
        this.platform = platform;
    }

    public FirebaseMessage(WhoAmI who, String displayName, String uid, String message) {
        this(who, displayName, uid, message, java.util.UUID.randomUUID().toString(), 0, Platform.ANDROID);
    }

    public static FirebaseMessage kiritan(String message) {
        return new FirebaseMessage(WhoAmI.KIRITAN, "", "", message, java.util.UUID.randomUUID().toString(), System.currentTimeMillis(), Platform.ANDROID);
    }

    static Optional<FirebaseMessage> fromSnapshot(DataSnapshot snapshot) {
        try {
            Log.i(TAG, snapshot.getKey() + " " + snapshot.getValue());
            int who = Integer.parseInt((String) snapshot.child(WHO).getValue());
            String displayName = (String)snapshot.child(DISPLAY_NAME).getValue();
            String uid = (String)snapshot.child(UID).getValue();
            String message = (String)snapshot.child(MESSAGE).getValue();
            String uuid = (String)snapshot.child(UUID).getValue();
            long createdAt = (long)snapshot.child(CREATED_AT).getValue();
            int platform = Integer.parseInt((String)snapshot.child(PLATFORM).getValue());
            return Optional.of(new FirebaseMessage(WhoAmI.values()[who], displayName, uid, message, uuid, createdAt, Platform.values()[platform]));
        } catch (Exception e) {
            Log.e(TAG, "from snapshot error", e);
            return Optional.absent();
        }
    }

    Map<String, Object> toMap() {
        HashMap<String, Object> m = new HashMap<>();
        m.put(WHO, String.valueOf(who.ordinal()));
        m.put(DISPLAY_NAME, displayName);
        m.put(UID, uid);
        m.put(MESSAGE, message);
        m.put(UUID, uuid);
        m.put(CREATED_AT, ServerValue.TIMESTAMP);
        m.put(PLATFORM, String.valueOf(platform.ordinal()));
        return m;
    }

    @Override
    public String toString() {
        return new JSONObject(toMap()).toString();
    }

    public WhoAmI getWho() {
        return who;
    }

    public String getDisplayName() {
        return displayName;
    }

    public String getUid() {
        return uid;
    }

    public String getMessage() {
        return message;
    }

    public String getUuid() {
        return uuid;
    }

    public long getCreatedAt() {
        return createdAt;
    }
}

File: slackbridge/src/main/scala/vrch/slackbridge/firebase/FirebaseMessage.scala
package vrch.slackbridge.firebase

import com.google.firebase.database.{DataSnapshot, ServerValue}
import vrch.slackbridge.firebase.FirebaseMessage._

import scala.collection.JavaConverters._

case class FirebaseMessage(who: WhoAmI,
                           uid: String,
                           displayName: String,
                           message: String,
                           uuid: String,
                           platform: Platform,
                           createdAt: Long) {

  def toMap: java.util.Map[String, Any] = Map[String, Any](
    WHO -> who.value.toString,
    UID -> uid,
    DISPLAY_NAME -> displayName,
    MESSAGE -> message,
    UUID -> uuid,
    PLATFORM -> platform.value.toString,
    CREATED_AT -> ServerValue.TIMESTAMP
  ).asJava
}


object FirebaseMessage {
  private val WHO = "who"
  private val DISPLAY_NAME = "display_name"
  private val UID = "uid"
  private val MESSAGE = "message"
  private val UUID = "uuid"
  private[firebase] val CREATED_AT = "created_at"
  private val PLATFORM = "platform"

  def apply(snapshot: DataSnapshot): FirebaseMessage = {
    FirebaseMessage(
      who = snapshot.asEnum(WHO, WhoAmI.values),
      uid = snapshot.asString(UID),
      displayName = snapshot.asString(DISPLAY_NAME),
      message = snapshot.asString(MESSAGE),
      uuid = snapshot.asString(UUID),
      platform = snapshot.asEnum(PLATFORM, Platform.values),
      createdAt = snapshot.asLong(CREATED_AT)
    )
  }

  def self(message: String): FirebaseMessage = {
    FirebaseMessage(
      who = WhoAmI.Self,
      uid = "",
      displayName = "anonymous@Slack",
      message = message,
      uuid = java.util.UUID.randomUUID().toString,
      platform = Platform.Slack,
      createdAt = 0
    )
  }

  def kiritan(message: String): FirebaseMessage = {
    FirebaseMessage(
      who = WhoAmI.Kiritan,
      uid = "",
      displayName = "",
      message = message,
      uuid = java.util.UUID.randomUUID().toString,
      platform = Platform.Slack,
      createdAt = 0
    )
  }
}
